File: android_java_hummus/src/main/java/com/github/klee0kai/hummus/adapterdelegates/delegate/simple/ClassAdapterDelegate.java
package com.github.klee0kai.hummus.adapterdelegates.delegate.simple;

import android.view.LayoutInflater;
import android.view.ViewGroup;

import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.hannesdorfmann.adapterdelegates4.AdapterDelegate;

import java.util.List;

public abstract class ClassAdapterDelegate<T> extends AdapterDelegate<List<Object>> {

    @LayoutRes
    private final int layoutRes;
    private final Class<T> tClass;
    private final boolean forceTheSameClass;

    public ClassAdapterDelegate(Class<T> tClass, @LayoutRes int layoutRes, boolean forceTheSameClass) {
        this.tClass = tClass;
        this.layoutRes = layoutRes;
        this.forceTheSameClass = forceTheSameClass;
    }

    public ClassAdapterDelegate(Class<T> tClass, @LayoutRes int layoutRes) {
        this(tClass, layoutRes, false);
    }

    @Override
    protected boolean isForViewType(@NonNull List<Object> items, int position) {
        Object item = items.get(position);
        if (forceTheSameClass) {
            return tClass.equals(item.getClass());
        } else {
            return tClass.isInstance(item);
        }
    }

    @NonNull
    @Override
    protected SimpleViewHolder onCreateViewHolder(ViewGroup parent) {
        return new SimpleViewHolder(LayoutInflater.from(parent.getContext()).inflate(layoutRes, parent, false));
    }

    @Override
    protected void onBindViewHolder(@NonNull List<Object> objects, int i, @NonNull RecyclerView.ViewHolder viewHolder, @NonNull List<Object> list) {
        onBindViewHolder((T) objects.get(i), i, (SimpleViewHolder) viewHolder);
    }


    protected abstract void onBindViewHolder(T it, int pos, @NonNull SimpleViewHolder vh);

}

File: android_kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/adapterdelegates/delegate/simple/ClassAdapterDelegate.kt
package com.github.klee0kai.hummus.adapterdelegates.delegate.simple

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.annotation.LayoutRes
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.hannesdorfmann.adapterdelegates4.AdapterDelegate


abstract class ClassAdapterDelegate<T>(
    protected open val tClass: Class<T>,
    @field:LayoutRes @param:LayoutRes protected open val layoutRes: Int,
    protected open val forceTheSameClass: Boolean = false
) : AdapterDelegate<List<Any>>() {

    override fun isForViewType(items: List<Any>, position: Int): Boolean {
        val item = items[position]
        return if (forceTheSameClass) {
            tClass == item.javaClass
        } else {
            tClass.isInstance(item)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup): SimpleViewHolder {
        return SimpleViewHolder(
            LayoutInflater.from(parent.context).inflate(
                layoutRes, parent, false
            )
        )
    }

    override fun onBindViewHolder(
        objects: List<Any>,
        pos: Int,
        viewHolder: ViewHolder,
        list: List<Any>
    ) {
        onBindViewHolder(objects[pos] as T, pos, viewHolder as SimpleViewHolder)
    }

    protected abstract fun onBindViewHolder(it: T, pos: Int, vh: SimpleViewHolder)


    companion object {

        fun <T> create(
            tClass: Class<T>,
            @LayoutRes layoutRes: Int,
            forceTheSameClass: Boolean = false,
            bind: SimpleViewHolder.(it: T, pos: Int) -> Unit
        ): ClassAdapterDelegate<T> =
            object : ClassAdapterDelegate<T>(tClass, layoutRes, forceTheSameClass) {
                override fun onBindViewHolder(it: T, pos: Int, vh: SimpleViewHolder) {
                    bind.invoke(vh, it, pos)
                }
            }

    }
}
